### DISCUSSION TITLE
Vertex Block Descent

### DISCUSSION THREAD

**Discussion by RT2Code** [jrouwe/JoltPhysics#1123] [General]
Hello,

A new promising physics solver was introduced recently in the latest siggraph conference: https://ankachan.github.io/Projects/VertexBlockDescent/index.html

What do you think about it? I suppose that replacing XPBD with it in Jolt would require a tremendous amount of work, but do you think it is something that could be done in the future?

**Comment by jrouwe** [comment]
Hello,

I am aware of this paper. I need to do some investigation before I can tell if this would be an improvement for Jolt or not. I think it does indeed mean a substantial refactor. 

As far as I can see now, the way the VBD solver works is really beneficial for GPU, but since Jolt currently runs on CPU it has much less problems with the 'vertex coloring' issue as mentioned in the paper. Also, I think the solver loop requires calculating some quantities per vertex instead of per constraint which XPBD does (edge constraint = 2 vertices, volume constraint = 4 vertices), which again is fine on GPU but less so on CPU.

**Comment by AnuraDev** [comment]
bump. right now i am searching for a solution to simulate simple cloth which should be purely visual and scale decently well for number of instances, some compute shader solution seems like best for this, cpu solutions are imo too expensive just for visual stuff like this. (ex. rts game with units who wear capes, collision should be probably just with the primitive shape of unit) right now i am not sure what something lightweight/performant like this exists (at least open source).
so the "question" is, are there any plans for gpu soft body simulation in jolt? (the gapi probably will be "problematic" but hlsl should be ok cause you can port it relative easily)

**Reply by jrouwe** [reply]
Yes, eventually I would like this (and other parts of the simulation) to be able to run on GPU. But this is a lot of work so will not land any time soon.

**Reply by AnuraDev** [reply]
yeah thats obvious, but good what it's on the roadmap. jolt ftw

**Comment by sirisian** [comment]
@jrouwe https://graphics.cs.utah.edu/research/projects/avbd/ There's this paper also on "Augmented Vertex Block Descent" that looks very impressive.

**Reply by jrouwe** [reply]
Thanks for mentioning that! I hadn't seen it yet. Looks very impressive indeed.

**Reply by Pombal** [reply]
Siggraph demo at 1h40min: https://www.youtube.com/live/ucRukZM0d1s?si=L8VLX3aipqwurDxo

**Reply by tomsterBG** [reply]
"There are two microphones on both sides of the room, yell at them to control the ball". Lovely demo!

**Reply by TheDiakou** [reply]
@jrouwe I know you've said in the past when you find the time to read in detail you'll check things out when it comes to AVBD, was just curious if you've formulated some initial thoughts or if still caught up in busy land! (truly just curious)

**Reply by jrouwe** [reply]
No, I have been on holidays and I've been really busy with a new feature for Jolt

**Reply by TheDiakou** [reply]
happy holidays! hope you had a great time. Jolt as a library is truly exciting. I can't wait to share our team's deterministic rollback netcode for unreal engine 5.6 relying on jolt. Its performance is genuinely wild and we've been able to make an incredibly smooth netcode around it for an "ARPG mobafighter" (kind of like battlerite)

**Reply by Pombal** [reply]
@TheDiakou Would you be willing to share more about your deterministic rollback implementation? We're starting to implement the same in our custom engine

**Reply by sirisian** [reply]
@jrouwe https://github.com/savant117/avbd-demo3d Forgot to mention this, but they uploaded this 3D version and included a 3D demo on the project page.

**Comment by addmix** [comment]
AVBD would be amazing. I just hope Godot can implement more of Jolt's features.
